### ISSUE TITLE
How was the rosenbridge coprocessor, and its instruction set, discovered?

### ISSUE THREAD

**Issue by i336** [xoreaxeaxeax/rosenbridge#6]
I think this is the penultimate question on everyone's minds right now.

Okay, so sandsifter did the heavy lifting; but that just firehoses random instructions at the CPU and looks for interesting results. That sounds like it would generate a nontrivial amount of noise - not so much so as to make the whole process overly tedious or simply unviable, but enough that skill and experience would be paramount to figure out what's worth following up on and what can be safely discarded.

But sandsifter just deals in discovery, and (in this case) to a very rudimentary extent - sandsifter was designed to find oddness in x86 CPUs, not entirely new processor architectures!

I am ***very*** interested to find out how you went from "hmm, that's weird" to pinpointing/establishing the very *existence* of the coprocessor, to identifying ***21*** of its instructions - and the x86 wrappings!

Also, in the same way a fighter pilot might share how to start up and fly an F-18, would you mind sharing how work on the assembler could be furthered by anyone with a VIA C3 who's interested in playing with this beyond going "huh, it's vulnerable"?

Please tell us that research papers and/or in-depth blog posts are in the pipeline. :)

And thanks, too. This is really awesome. ME, eat your heart out.

**Comment by szymonmatuszczak** [comment]
Yeah. That's why I don't think the Qubes OS is an ultra-secure OS as Rutkowska says, at least on a modern x86 hardware (new versions rely on VT-x and VT-d, so i can't use it on a non-modern x86 hardware). Thinking in the same direction, QubesOS' movement from Xen PV to VT-x could be a step backwards, because we can easily inspect how Xen PV works, but inspecting VT-x (on firmware and circuitry) is very hard, but using PV, a JavaScript malware loaded in a webbrowsing qube utilizing a hidden backdoor still would probably be able to switch itself with a magic way to ring 0/ring -1, read some data from memory/any DMA device and send it back in a next HTTP request... so if a CPU is backdoored, utilising or not utilising VT-x doesn't really seem to change anything. If I would want a really secure computing device, I could port Qubes OS to some chinese-manufactured in-order-executing ARM, MIPS, PowerPC or even a 68060 (which are probably less likely to be backdoored, especially the last one)... but they still can be backdoored. 68060 are enough simple to be inspected with a bowl of acid and a microscope, but still it's complicated. Well, maybe RISC-V would be secure enough, it would be very hard to backdoor one implemented in a FPGA or even synthesized and ordered as an ASIC in some Chinese factory.

**Comment by brundage** [comment]
[This is how I found my way here](https://www.tomshardware.com/news/x86-hidden-god-mode,37582.html).

Looks like it all started with a patent search.

**Comment by ajxs** [comment]
In this video Mr Domas details his methodology and answers some of your questions above: https://youtu.be/_eSAF_qT_FY
It has been pointed out elsewhere that the DEC(sic) has been documented elsewhere in leaked proprietary documents. An online search for `via C3 ALTINST` should yield some useful results.
Not to detract in any way from Mr Domas' research, the video presentation detailing his methodology is simply amazing.
